internal/icinga2/client.go

```diff
@@ -0,0 +1,388 @@
+		Name:      eventName,
+		URL:       eventUrl.String(),
+		Tags:      eventTags,
+		ExtraTags: eventExtraTags,
+	}, nil
+}
+
+// buildHostServiceEvent constructs an event.Event based on a CheckResult, a Host or Service state, a Host name and an
+// optional Service name if the Event should represent a Service object.
+func (client *Client) buildHostServiceEvent(ctx context.Context, result CheckResult, state int, host, service string) (*event.Event, error) {
+	var eventSeverity event.Severity
+
+	if service != "" {
+		switch state {
+		case StateServiceOk:
+			eventSeverity = event.SeverityOK
+		case StateServiceWarning:
+			eventSeverity = event.SeverityWarning
+		case StateServiceCritical:
+			eventSeverity = event.SeverityCrit
+		default: // UNKNOWN or faulty
+			eventSeverity = event.SeverityErr
+		}
+	} else {
+		switch state {
+		case StateHostUp:
+			eventSeverity = event.SeverityOK
+		case StateHostDown:
+			eventSeverity = event.SeverityCrit
+		default: // faulty
+			eventSeverity = event.SeverityErr
+		}
+	}
+
+	ev, err := client.buildCommonEvent(ctx, host, service)
+	if err != nil {
+		return nil, err
+	}
+
+	ev.Type = event.TypeState
+	ev.Severity = eventSeverity
+	ev.Message = result.Output
+
+	return ev, nil
+}
+
+// buildAcknowledgementEvent from the given fields.
+func (client *Client) buildAcknowledgementEvent(ctx context.Context, host, service, author, comment string) (*event.Event, error) {
+	ev, err := client.buildCommonEvent(ctx, host, service)
+	if err != nil {
+		return nil, err
+	}
+
+	ev.Type = event.TypeAcknowledgement
+	ev.Username = author
+	ev.Message = comment
+
+	return ev, nil
+}
+
+// startCatchupWorkers launches goroutines for catching up the Icinga 2 API state.
+//
+// Each event will be sent to the returned channel. When all launched workers have finished - either because all are
+// done or one has failed and the others were interrupted -, the channel will be closed. In case of a failure, _one_
+// final error will be sent back.
+//
+// Those workers honor a context derived from the Client.Ctx and would either stop when this context is done or when the
+// context.CancelFunc is called.
+//
+// The startup time might be delayed through the parameter. This lets the goroutines sleep to rate-limit reconnection
+// attempts during network hiccups.
+func (client *Client) startCatchupWorkers(delay time.Duration) (chan *catchupEventMsg, context.CancelFunc) {
+	startTime := time.Now()
+	catchupEventCh := make(chan *catchupEventMsg)
+
+	// Unfortunately, the errgroup context is hidden, that's why another context is necessary.
+	ctx, cancel := context.WithCancel(client.Ctx)
+	group, groupCtx := errgroup.WithContext(ctx)
+
+	objTypes := []string{"host", "service"}
+	for _, objType := range objTypes {
+		objType := objType // https://go.dev/doc/faq#closures_and_goroutines
+		group.Go(func() error {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-time.After(delay):
+			}
```

Comment:
Is the event loop in `client.worker()` expected to do anything while these are sleeping? Otherwise, wouldn't it make more sense to sleep in that loop before calling this function?

Reply:
I don't want to block the `worker`'s event loop with a busy sleep. Otherwise, all other signals have to wait, including the cancellation context. Thus, by moving the sleep within the `startCatchupWorkers` worker Goroutines, they don't block anything and can easily be canceled through their custom context.

Yes, I am not so proud of this design, but it has the least implications.